core/src/main/cpp/gdn/projection_mesh/projection_mesh.h

```diff
@@ -0,0 +1,213 @@
+        return this->collision_shape_ref;
+    }
+
+    inline void set_external_texture(Ref<ExternalTexture> external_texture) {
+        if (!shader_material_ref.is_valid()) {
+            ALOGE("Cannot set ExternalTexture, invalid ShaderMaterial.");
+            return;
+        }
+        shader_material_ref->set_shader_param(kGastTextureParamName, external_texture);
+    }
+
+    inline ExternalTexture* get_external_texture() {
+        if (!shader_material_ref.is_valid()) {
+            return nullptr;
+        }
+
+        ExternalTexture *external_texture = nullptr;
+        Ref<Texture> texture = shader_material_ref->get_shader_param(kGastTextureParamName);
+        if (texture.is_valid()) {
+            external_texture = Object::cast_to<ExternalTexture>(*texture);
+        }
+        return external_texture;
+    }
+
+    inline void set_render_on_top(bool enable) {
+        if (this->render_on_top == enable) {
+            return;
+        }
+        this->render_on_top = enable;
+        if (shader_material_ref.is_valid() && shader_material_ref->get_shader().is_valid()) {
+            String shader_code = shader_material_ref->get_shader()->get_code();
+            shader_code += kDisableDepthTestRenderMode;
+            shader_material_ref->get_shader()->set_code(shader_code);
+        }
+        update_render_priority();
+    }
+
+    inline bool is_render_on_top() {
+        return this->render_on_top;
+    }
+
+    inline void set_gaze_tracking(bool gaze_tracking) {
+        if (this->gaze_tracking == gaze_tracking) {
+            return;
+        }
+        this->gaze_tracking = gaze_tracking;
+        update_render_priority();
+        if (shader_material_ref.is_valid()) {
+            shader_material_ref->set_shader_param(kGastEnableBillBoardParamName, gaze_tracking);
+        }
+    }
+
+    inline bool is_gaze_tracking() {
+        return gaze_tracking;
+    }
+
+    inline void update_render_priority() {
+        ShaderMaterial *shader_material = *shader_material_ref;
+        if (!shader_material) {
+            return;
+        }
+
+        int render_priority = 0;
+        if (render_on_top) {
+            render_priority++;
+            if (gaze_tracking) {
+                render_priority++;
+            }
+        }
+        shader_material->set_render_priority(render_priority);
+    }
+
+    inline void set_alpha(float alpha) {
+        if (this->alpha == alpha) {
+            return;
+        }
+        this->alpha = alpha;
+        shader_material_ref->set_shader_param(kGastNodeAlphaParamName, alpha);
+    }
+
+    inline StereoMode get_stereo_mode() {
+        return stereo_mode;
+    }
+
+    inline void set_stereo_mode() {
+        // TODO: Implement
+    }
+
+    inline void reset_mesh() {
+        mesh_instance->set_mesh(Ref<Resource>());
+        collision_shape_ref = Ref<Resource>();
+    }
+
+    inline void reset_external_texture() {
+        set_external_texture(Ref<Resource>());
+    }
```

Comment:
Should these methods be called in the destructor?

Reply:
Done.